File: core/SubstringWithConcatenationOfAllWords/SubstringWithConcatenationOfAllWords.java
package com.dbc.code;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SubstringWithConcatenationOfAllWords {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        Map<String, Integer> map = new HashMap<>();
        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }
        int index = 0;
        while (index <= s.length() - words.length * words[0].length()) {
            Map<String, Integer> tempMap = new HashMap<>(map);
            boolean flag = true;
            for (int point = index; point < index + words.length * words[0].length(); point+=words[0].length()) {
                String sub = s.substring(point, point + words[0].length());
                if (!tempMap.containsKey(sub) || tempMap.get(sub) <= 0) {
                    flag = false;
                    break;
                }
                tempMap.put(sub, tempMap.get(sub) - 1);
            }
            if (flag) {
                res.add(index);
            }
            index++;
        }
        return res;
    }
}

File: core/SubstringWithConcatenationOfAllWords/SubstringWithConcatenationOfAllWords.py
from typing import List
from copy import deepcopy


def findSubstring(s: str, words: List[str]) -> List[int]:
    word_len, size, length = len(words[0]), len(words), len(s)
    index, word_map, res = 0, {}, []
    for word in words:
        word_map[word] = word_map.get(word, 0) + 1
    while index <= length - size * word_len:
        temp_map, flag = deepcopy(word_map), True
        for point in range(index, index + size * word_len, word_len):
            if s[point:point+word_len] not in temp_map or temp_map[s[point:point+word_len]] <= 0:
                flag = False
                break
            temp_map[s[point:point+word_len]] -= 1

        if flag:
            res.append(index)

        index += 1

    return res





